[JIT] Add SchemaInfo python bindings to init.cpp

ttanpcs opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #81655
* #81653
* #81562
* #81561
* #81575
* __->__ #81518

- Added python bindings for SchemaInfo class, SchemaArgument struct, and SchemaArgType enum. 
- Tested that argument values are added correctly to SchemaInfo binding in test_schema_check.py

Commit:
[JIT] Add SchemaInfo python bindings to init.cpp

[ghstack-poisoned]

Commit:
Update on "[JIT] Add SchemaInfo python bindings to init.cpp"




[ghstack-poisoned]

Commit:
Update on "[JIT] Add SchemaInfo python bindings to init.cpp"




[ghstack-poisoned]

davidberard98 commented on test/jit/test_schema_check.py:
```diff
@@ -342,3 +342,41 @@ def test_is_alias_of(self):
+        schemaInfoCheck = SchemaInfoBindTestMode()
+        with enable_torch_dispatch_mode(schemaInfoCheck):
+            x.add(x)
+        self.assertEqual([False, False, True, True], schemaInfoCheck.check)
```
nit: since the dispatch mode already seems like it's pretty closely tied into the `x.add(x)` test, it might be clearer if you just put the asserts directly in the torch_dispatch mode

davidberard98 commented on torch/csrc/jit/python/init.cpp:
```diff
@@ -1638,7 +1640,63 @@ void initJITBindings(PyObject* module) {
+          [](SchemaInfo& self,
+             const std::string& name,
+             const py::object& value) {
+            if (name == "input") {
+              self.addArgumentValue("self", toTypeInferredIValue(value));
```
Would it be okay to leave this logic on the python side since you already have this logic in schema check mode? Or is there a reason we need it here as well?

goldenxuett replied:
It's necessary here due to how 'normalize_function' ends up naming the first argument when constructing a dictionary of argument names to argument values. Regardless of what the function's first argument name is supposed to be, 'normalize_function' will make the argument name "input." 

To match this, the standardize_name() function on the python side changes all names of "self" into "input." 

However, in this add_argument_value, we need the original name of the argument, not the normalized name which is always "input," thus this check exists here so that extra processing doesn't need to occur on the python side.

Commit:
Update on "[JIT] Add SchemaInfo python bindings to init.cpp"


- Added python bindings for SchemaInfo class, SchemaArgument struct, and SchemaArgType enum. 
- Tested that argument values are added correctly to SchemaInfo binding in test_schema_check.py

[ghstack-poisoned]

Commit:
Update on "[JIT] Add SchemaInfo python bindings to init.cpp"


- Added python bindings for SchemaInfo class, SchemaArgument struct, and SchemaArgType enum. 
- Tested that argument values are added correctly to SchemaInfo binding in test_schema_check.py

[ghstack-poisoned]

davidberard98 reviewed (approved):
nice!
